File: app/src/core/scheduler.ts
/**
 * Hebrew-to-Gregorian date scheduling — ported from Python scheduler.py.
 *
 * Uses @hebcal/core's HDate for Hebrew ↔ Gregorian conversion.
 */

import { HDate } from "@hebcal/core";
import type { ScheduledEvent } from "./types";

// ---------------------------------------------------------------------------
// Month-key → @hebcal/core month name mapping
// ---------------------------------------------------------------------------

const MONTH_KEY_TO_HEBCAL: Record<string, string> = {
  tishrei: "Tishrei",
  cheshvan: "Cheshvan",
  kislev: "Kislev",
  teves: "Tevet",
  shevat: "Sh'vat",
  adar: "Adar",
  adar1: "Adar I",
  adar2: "Adar II",
  nissan: "Nisan",
  iyar: "Iyyar",
  sivan: "Sivan",
  tammuz: "Tamuz",
  av: "Av",
  elul: "Elul",
};

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/**
 * Resolve the hebcal month name for a given monthKey and Hebrew year,
 * applying the standard Adar / leap-year rules:
 *
 * - "adar" in a leap year  → Adar II  (the "main" Adar for birthdays etc.)
 * - "adar2" in a non-leap year → Adar  (fall back to the only Adar)
 */
function hebcalMonthName(monthKey: string, hebrewYear: number): string {
  const base = MONTH_KEY_TO_HEBCAL[monthKey];
  if (base === undefined) {
    throw new Error(`Unsupported month key '${monthKey}'.`);
  }

  const leap = HDate.isLeapYear(hebrewYear);

  if (monthKey === "adar" && leap) {
    return "Adar II";
  }
  if (monthKey === "adar2" && !leap) {
    return "Adar";
  }

  return base;
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/**
 * Generate the next `count` Gregorian occurrences of a recurring Hebrew date,
 * starting no earlier than `fromDate`.
 *
 * Returns events with an empty `summary` — the caller is responsible for
 * filling it in (mirrors the Python pipeline pattern).
 */
export function generateOccurrences(
  day: number,
  monthKey: string,
  fromDate: Date,
  count: number,
): ScheduledEvent[] {
  if (count < 1) {
    throw new RangeError("Occurrence count must be at least 1.");
  }

  const startHDate = new HDate(fromDate);
  let hebrewYear = startHDate.getFullYear();

  const occurrences: ScheduledEvent[] = [];

  while (occurrences.length < count) {
    const month = hebcalMonthName(monthKey, hebrewYear);
    const hd = new HDate(day, month, hebrewYear);
    const greg = hd.greg();

    if (greg >= fromDate) {
      occurrences.push({
        summary: "",
        startDate: greg,
        hebrewDateText: `${day} ${month} ${hebrewYear}`,
      });
    }

    hebrewYear += 1;
  }

  return occurrences;
}

/**
 * Convenience wrapper: generate `yearsAhead` future occurrences of a Hebrew
 * date and stamp each event with the given title.
 */
export function generateEvents(
  title: string,
  day: number,
  monthKey: string,
  yearsAhead: number,
): ScheduledEvent[] {
  const events = generateOccurrences(day, monthKey, new Date(), yearsAhead);

  for (const event of events) {
    event.summary = title;
  }

  return events;
}

File: hebrew_dates_to_cal/scheduler.py
from __future__ import annotations

from datetime import date

from pyluach import dates
from pyluach.hebrewcal import Year

from .models import EventInput, HebrewDateSpec, ScheduledEvent

FIXED_MONTH_NUMBERS = {
    "nissan": 1,
    "iyar": 2,
    "sivan": 3,
    "tammuz": 4,
    "av": 5,
    "elul": 6,
    "tishrei": 7,
    "cheshvan": 8,
    "kislev": 9,
    "teves": 10,
    "shevat": 11,
}


def _month_number_for_year(month_key: str, hebrew_year: int) -> int:
    if month_key in FIXED_MONTH_NUMBERS:
        return FIXED_MONTH_NUMBERS[month_key]
    is_leap = Year(hebrew_year).leap
    if month_key == "adar":
        return 13 if is_leap else 12
    if month_key == "adar1":
        return 12
    if month_key == "adar2":
        return 13 if is_leap else 12
    raise ValueError(f"Unsupported month key '{month_key}'.")


def _gregorian_for_hebrew_year(spec: HebrewDateSpec, hebrew_year: int) -> date:
    month_number = _month_number_for_year(spec.month_key, hebrew_year)
    try:
        gregorian = dates.HebrewDate(hebrew_year, month_number, spec.day).to_greg()
    except ValueError as exc:
        raise ValueError(
            f"Hebrew date '{spec.original_text}' does not exist in Hebrew year {hebrew_year}."
        ) from exc
    return date(gregorian.year, gregorian.month, gregorian.day)


def generate_occurrences(spec: HebrewDateSpec, from_date: date, count: int) -> list[date]:
    if count < 1:
        raise ValueError("Occurrence count must be at least 1.")
    current_hebrew = dates.GregorianDate(from_date.year, from_date.month, from_date.day).to_heb()
    occurrences: list[date] = []
    hebrew_year = current_hebrew.year
    while len(occurrences) < count:
        candidate = _gregorian_for_hebrew_year(spec, hebrew_year)
        if candidate >= from_date:
            occurrences.append(candidate)
        hebrew_year += 1
    return occurrences


def build_scheduled_events(
    events: list[EventInput],
    *,
    from_date: date,
    count: int,
) -> list[ScheduledEvent]:
    scheduled: list[ScheduledEvent] = []
    for event in events:
        summary = event.render_summary()
        for occurrence in generate_occurrences(event.hebrew_date, from_date=from_date, count=count):
            scheduled.append(
                ScheduledEvent(
                    summary=summary,
                    start_date=occurrence,
                    hebrew_date_text=event.hebrew_date.original_text,
                )
            )
    scheduled.sort(key=lambda item: (item.start_date, item.summary))
    return scheduled
